syntax = "proto3";

package user;

service UserCommandService {
  rpc RegisterUser (RegisterUserRequest) returns (RegisterUserResponse);
  rpc UpdateUser (UpdateUserRequest) returns (UpdateUserResponse);
  rpc UpdateValue (UpdateValueRequest) returns (UpdateValueResponse);
  rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
  rpc DeleteDataByTime (DeleteDataByTimeRequest) returns (DeleteDataByTimeResponse);
}

service UserQueryService {
  rpc GetAllData (Empty) returns (AllDataResponse);
  rpc GetLastStockValue (EmailRequest) returns (StockValueResponse);
  rpc GetAverageStockValue (AverageStockRequest) returns (StockValueResponse);
}

message RegisterUserRequest {
  string email = 1;
  string ticker = 2;
  float low_value = 3;
  float high_value = 4;
}

message RegisterUserResponse {
  string message = 1;
}

message UpdateUserRequest {
  string email = 1;
  string ticker = 2;
  float low_value = 3;
  float high_value = 4;
}

message UpdateUserResponse {
  string message = 1;
}

message UpdateValueRequest {
  string email = 1;
  float low_value = 2;
  float high_value = 3;
}

message UpdateValueResponse {
  string message = 1;
}

message DeleteUserRequest {
  string email = 1;
}

message DeleteUserResponse {
  string message = 1;
}

message Empty {}

message EmailRequest {
  string email = 1;
}

message AverageStockRequest {
  string email = 1;
  int32 count = 2;
}

message StockValueResponse {
  string message = 1;
  float value = 2;
}

message AllDataResponse {
  repeated string data = 1;
}

message DeleteDataByTimeRequest {
  int64 start_time = 1;
}

message DeleteDataByTimeResponse {
  string message = 1;
}